### PR TITLE
A click on Overdue Assignment card filters all records on the dashboard

### PR DESCRIPTION
**Pull request by Vitalii-Misechko**
[US119013](https://rally1.rallydev.com/#/detail/userstory/408910753872?fdp=true): [Engagement] Interactions - Assignments

ACs: 
- [x] Clicking on the value in the Overdue Assignments summary card filters the returned data for just the users and courses where the users have 1 or more overdue assignments.
- [x] Summarize the interaction and filter applied with the ability to "X" out to remove and revert back.
- [x] If another interaction occurs before or after this one, these actions are combined.

### Functional Testing 
* Click on the number of Overdue Assignment card. 
  * Ensure that Engagement Dashboard shows filtered data
  * Ensure that Applied Filters - panel appeared
  * Ensure that Applied Filters contains Clear All and Overdue Assignment buttons
* Click on the number of Overdue Assignment card again. Ensure nothing changed
* A click on Clear All button of Applied Filters clears  Overdue Assignment card filter and hides Applied Filters panel
* A click on Overdue Assignment button of Applied Filters clears  Overdue Assignment card filter and hides Applied Filters panel
* Applied Filters is accessible with Tab key

### UX and docs - N/A

- [x] Unit tests
- [x] Manual tests

### CODE REVIEW

**Comment by Vitalii-Misechko** [comment]
image

**Comment by Vitalii-Misechko** [comment]
The clear button is still in progress.

File: components/overdue-assignments-card.js [hunk 1]
```diff
@@ -31,8 +31,24 @@ class OverdueAssignmentsCard extends Localizer(MobxLitElement) {
 
 	render() {
 		return html`
-			<d2l-labs-summary-card is-value-clickable id="d2l-insights-engagement-overdue-assignments" .data="${this.data}" card-title="${this._cardTitle}" card-value="${this._cardValue}" card-message="${this._cardMessage}"></d2l-labs-summary-card>
+			<d2l-labs-summary-card
+				is-value-clickable id="d2l-insights-engagement-overdue-assignments"
```
Comment (johngwilkinson) [inline components/overdue-assignments-card.js:35]: nit: I'd prefer if `id=...` were on its own line

File: components/overdue-assignments-card.js [hunk 2]
```diff
@@ -31,8 +31,24 @@ class OverdueAssignmentsCard extends Localizer(MobxLitElement) {
...
 		`;
 	}
+
+	_valueClickHandler() {
+		// Temp code to toggle filter
+		const isApplied = this.data.cardFilters['d2l-insights-overdue-assignments-card'].isApplied;
+		console.log(`Is applied: ${isApplied}.`);
+		this.data.setApplied('d2l-insights-overdue-assignments-card', !isApplied);
+
+		//this.data.setApplied('d2l-insights-overdue-assignments-card', true); // production code
```
Comment (johngwilkinson) [inline components/overdue-assignments-card.js:51]: Can this be cleaned up now?

File: components/results-card.js [hunk 3]
```diff
@@ -1,38 +1,38 @@
...
+	get _cardMessage() {
+		return this.localize('components.insights-engagement-dashboard.resultsReturned');
+	}
+
+	get _cardTitle() {
+		return this.localize('components.insights-engagement-dashboard.resultsHeading');
+	}
+
+	get _cardValue() {
+		return this.data.records.length;
```
Comment (johngwilkinson) [inline components/results-card.js:29]: Why this change? Isn't this card supposed to show the number of users in view, with all filters applied?

File: engagement-dashboard.js [hunk 4]
```diff
@@ -82,6 +82,11 @@ class EngagementDashboard extends Localizer(LitElement) {
 				// 	countUniqueField: 'UserId',
 				// 	messageProvider: () => 'users'
 				// }
+				{
+					id: 'd2l-insights-overdue-assignments-card',
+					title: 'Overdue Assignments',
+					filter: (record) => record[RECORD.OVERDUE] > 0
```
Comment (johngwilkinson) [inline engagement-dashboard.js:88]: No change needed, but it would be nice if this filter function were defined in the overdue filter component itself.

File: engagement-dashboard.js [hunk 5]
```diff
@@ -109,7 +114,7 @@ class EngagementDashboard extends Localizer(LitElement) {
 					<d2l-insights-results-card .data="${this._data}"></d2l-insights-results-card>
 					<d2l-insights-current-final-grade-card .data="${this._data}"></d2l-insights-current-final-grade-card>
 					<d2l-insights-overdue-assignments-card .data="${this._data}"></d2l-insights-overdue-assignments-card>
-					<d2l-insights-debug-card .data="${this._data}" metric-to-display="recordsLength" title="Records" message="number of records within filter parameters"></d2l-insights-debug-card>
+					<d2l-insights-debug-card .data="${this._data}" metric-to-display="filteredRecordsLength" title="Records" message="number of records within filter parameters"></d2l-insights-debug-card>
```
Comment (johngwilkinson) [inline engagement-dashboard.js:117]: We're pretty close to shipping something, so we should comment this out, I guess.

File: model/cardFilter.js [hunk 6]
```diff
@@ -11,6 +11,7 @@ export class CardFilter {
 		this.countUniqueField = countUniqueField;
 		this.data = data;
 		this.isApplied = isApplied;
+		this.filter = filter || this._defaultFilter;
```
Comment (johngwilkinson) [inline model/cardFilter.js:14]: you would need to `bind` `this._defaultFilter`, I think - not sure we need this, in any case

File: components/overdue-assignments-card.js (hunk 2 above)
Comment (Vitalii-Misechko) [reply components/overdue-assignments-card.js]: done

File: components/overdue-assignments-card.js [hunk 7]
```diff
@@ -31,8 +31,19 @@ class OverdueAssignmentsCard extends Localizer(MobxLitElement) {
...
+				card-title="${this._cardTitle}"
+				card-value="${this._cardValue}"
+				card-message="${this._cardMessage}"
+				@d2l-labs-summary-card-value-click=${this._valueClickHandler}
+			></d2l-labs-summary-card>
 		`;
 	}
+
+	_valueClickHandler() {
+		this.data.setApplied('d2l-insights-overdue-assignments-card', true);
```
Comment (johngwilkinson) [inline components/overdue-assignments-card.js:46]: Shouldn't a second click turn the filter off? I've been assuming that, but maybe I'm wrong.

Comment (Vitalii-Misechko) [reply components/overdue-assignments-card.js:46]: Hm...good question. I redirect it to Megan and Kevin

File: components/results-card.js (hunk 3 above)
Comment (Vitalii-Misechko) [reply components/results-card.js:29]: I thought it showed number of all records fetched from a server. I'll revert it back

Comment (Vitalii-Misechko) [reply components/results-card.js]: done

File: engagement-dashboard.js (hunk 5 above)
Comment (Vitalii-Misechko) [reply engagement-dashboard.js]: done

File: model/cardFilter.js (hunk 6 above)
Comment (Vitalii-Misechko) [reply model/cardFilter.js:14]: removed

File: components/overdue-assignments-card.js (hunk 7 above)
Comment (Vitalii-Misechko) [reply components/overdue-assignments-card.js:47]: Megan and Kevin approved it this way. Separate buttons are responsible for clearing filters

**Comment by Vitalii-Misechko** [comment]
New screenshot of filtered state
image

**Comment by Vitalii-Misechko** [comment]
Some Saucelabs problems

```
10 09 2020 17:36:34.357:ERROR [SaucelabsLauncher]: Error: 10 Sep 17:36:34 - Sauce Connect could not establish a connection.
10 Sep 17:36:34 - Please check your firewall and proxy settings.
    at Socket.<anonymous> (/home/travis/build/Brightspace/insights-engagement-dashboard/node_modules/saucelabs/build/index.js:312:25)
    at Socket.emit (events.js:314:20)
    at Socket.EventEmitter.emit (domain.js:486:12)
    at addChunk (_stream_readable.js:307:12)
    at readableAddChunk (_stream_readable.js:282:9)
    at Socket.Readable.push (_stream_readable.js:221:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:188:23)
```

**Comment by johngwilkinson** [comment]
> Some Saucelabs problems
> 
> ```
> 10 09 2020 17:36:34.357:ERROR [SaucelabsLauncher]: Error: 10 Sep 17:36:34 - Sauce Connect could not establish a connection.
> 10 Sep 17:36:34 - Please check your firewall and proxy settings.
>     at Socket.<anonymous> (/home/travis/build/Brightspace/insights-engagement-dashboard/node_modules/saucelabs/build/index.js:312:25)
>     at Socket.emit (events.js:314:20)
>     at Socket.EventEmitter.emit (domain.js:486:12)
>     at addChunk (_stream_readable.js:307:12)
>     at readableAddChunk (_stream_readable.js:282:9)
>     at Socket.Readable.push (_stream_readable.js:221:10)
>     at Pipe.onStreamRead (internal/stream_base_commons.js:188:23)
> ```

https://d2l.slack.com/archives/CMNGJ22G0/p1599758612147000

**Comment by Vitalii-Misechko** [comment]
JFYI 
@rohitvinnakota @hyehayes @anhill-D2L @mgalian

**Review by johngwilkinson — APPROVED** [review]

**Comment by johngwilkinson** [comment]
@bemailloux this looks good to me now (possibly with a follow-up to change the "Applied Filters" text). I'm looking into the Sauce issue, but we may just want to force a merge today so we can get BSI up-to-date.

**Comment by bemailloux** [comment]
> @bemailloux this looks good to me now (possibly with a follow-up to change the "Applied Filters" text). I'm looking into the Sauce issue, but we may just want to force a merge today so we can get BSI up-to-date.

This will make it so that only the Overdue assignments card can be interacted with. Do we want that functionality to be shipped by itself?

**Comment by johngwilkinson** [comment]
> > @bemailloux this looks good to me now (possibly with a follow-up to change the "Applied Filters" text). I'm looking into the Sauce issue, but we may just want to force a merge today so we can get BSI up-to-date.
> 
> This will make it so that only the Overdue assignments card can be interacted with. Do we want that functionality to be shipped by itself?

I think so, but I'll ask on Slack.

**Review by bemailloux — APPROVED** [review]

**Comment by bemailloux** [comment]
Something I thought of that's probably a future improvement: is it possible to tab through the cards and press Enter or space to apply filters? (If not there's a catch-all a11y improvements story that can probably go into)

**Comment by johngwilkinson** [comment]
> > > @bemailloux this looks good to me now (possibly with a follow-up to change the "Applied Filters" text). I'm looking into the Sauce issue, but we may just want to force a merge today so we can get BSI up-to-date.
> > 
> > 
> > This will make it so that only the Overdue assignments card can be interacted with. Do we want that functionality to be shipped by itself?
> 
> I think so, but I'll ask on Slack.

Confirmed.
